task/mgmt-gateway/src/update/mod.rs

```diff
@@ -0,0 +1,63 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::mgs_handler::BufferMutex;
+use gateway_messages::{
+    ComponentUpdatePrepare, ResponseError, UpdateId, UpdateStatus,
+};
+
+#[cfg(feature = "gimlet")]
+pub(crate) mod host_flash;
+
+mod common;
+pub(crate) mod sp;
+
+// TODO Currently we only have one implementor of this trait
+// (`HostFlashUpdate`), and we don't actually use the trait directly anywhere.
+// This is currently just a guide for how we expect component update
+// implementations to be structured; once we have more than one, we can adjust
+// this trait definition and/or see if it makes sense to have some generic
+// support code.
+//
+// `SpUpdate` has methods with similar names/signatures, but aren't quite the
+// same due to having to deal with both auxflash and SP images. It therefore
+// does not implement this trait. That's another wart that we might be able to
+// address once we start adding more implementors and tweak this trait.
+pub(crate) trait ComponentUpdater {
+    /// Size of one block / sector / page; whatever unit the underlying update
+    /// mechanism wants as a single chunk.
+    fn block_size(&self) -> usize;
+
+    /// Attempt to start preparing for an update, using `buffer` as the backing
+    /// store for incoming data.
+    fn prepare(
+        &mut self,
+        buffer: &'static BufferMutex,
+        update: ComponentUpdatePrepare,
+    ) -> Result<(), ResponseError>;
+
+    /// Returns true if this task needs `step_preparation()` called.
+    fn is_preparing(&self) -> bool;
+
+    /// Do a small amount of preparation work.
+    fn step_preparation(&mut self);
+
+    /// Status of this update.
+    fn status(&self) -> UpdateStatus;
+
+    /// Attempt to ingest a single update chunk from MGS.
+    fn ingest_chunk(
+        &mut self,
+        id: &UpdateId,
+        offset: u32,
+        data: &[u8],
+    ) -> Result<(), ResponseError>;
+
+    /// Abort the current update if it matches `id`.
+    ///
+    /// If no update is in progress, should return `Ok(())`; i.e., this should
+    /// only return an error if we attempted to abort the update with id `id`
+    /// and that abort failed.
```

Comment:
Add comment about happens if an update is in progress, but with a different id – is that `Ok(())` or an error?

Reply:
I think I'm coming around to your position. _Someone_ has to decide what to do if we try to abort a completed update, and with the implementation as it is now, we swallow that information, which would force MGS to make more status calls (all inherently racy) to figure out what's going on. I'll change this to:

* If `id != current.id`, return an error no matter what our state is.
* If the ID matches and we're in state `Complete`, return an error.
* If the ID matches and we're in an active state or `Aborted`, succeed and return `Ok(())`.

I'll make similar changes to the SP update abort.

Reply:
Changed in d70746d5